### ISSUE TITLE
create a libretro maintainers team?

### ISSUE THREAD

**Issue by aanderse** [NixOS/nixpkgs#146402]
We could benefit from a `libretro` maintainers team. `retroarch` is no longer the only piece of software utilizing `libretro` cores, and a number of people are willing to maintain the software so a team makes sense.

ping @thiagokokada from https://github.com/NixOS/nixpkgs/pull/146295#issuecomment-971685567

**Comment by thiagokokada** [comment]
Yeah, this would be good.

Our current `retroarch` derivation also has many issues:
- Can't upgrade to latest version because of an issue: https://github.com/NixOS/nixpkgs/pull/146295#issuecomment-971536194
- There is no core metadata (e.g.: https://github.com/libretro/libretro-super/blob/master/dist/info/mame_libretro.info). This is probably related to the issue above
- There is no assets by default (needs to download it using `Update Assets`, not really the way we do it on Nix)
- Before #146295 it was quite a while since the last bump of libretro cores (the last one was 2020-03-06)

**Comment by aanderse** [comment]
I started working on standardizing how cores are included in `retroarch` as well, using `withPackages` instead of config options. I thought this would be a good idea.

**Comment by thiagokokada** [comment]
> I started working on standardizing how cores are included in `retroarch` as well, using `withPackages` instead of config options. I thought this would be a good idea.

Nice. I was going to do it on #146295 too but since there is already so many changes happening on it I decided to not do and leave for another PR.

Gladly to know that someone else is doing it.

**Comment by thiagokokada** [comment]
>     * Can't upgrade to latest version because of an issue: [retroarch: 1.8.5 -> 1.9.2 #146295 (comment)](https://github.com/NixOS/nixpkgs/pull/146295#issuecomment-971536194)
> 
>     * There are no core metadata (e.g.: https://github.com/libretro/libretro-super/blob/master/dist/info/mame_libretro.info). This is probably related to the issue above

As I suspected, the lack of core metadata was the reason why newer versions of RetroArch doesn't work. PR https://github.com/NixOS/nixpkgs/pull/146577 fixes both of those issues, and allow RetroArch to proper work (on the menu now it even shows things like the license of the core and as such, much nicer than before).

After #146577, I will probably start to package some missing cores (the idea is not to package everything, but some interesting cores like [DuckStation](https://www.libretro.com/index.php/new-playstation1-core-duckstation-now-available-for-retroarch/) are missing right now) and maybe work in a proper module (so I can solve one issue that #146577 introduces)?

**Comment by bjornfor** [comment]
Last time I looked at retroarch I gave up using it with Nix since upstream (handling of and use of) retroach.cfg is a mess. I said to myself, that must be fixed upstream. After that we can improve the Nix expr. (I currently use the flatpak.) I'm happy to see interest here though!

**Comment by thiagokokada** [comment]
> Last time I looked at retroarch I gave up using it with Nix since upstream (handling of and use of) retroach.cfg is a mess. I said to myself, that must be fixed upstream. After that we can improve the Nix expr. (I currently use the flatpak.) I'm happy to see interest here though!

Can you explain or give some examples why handling and use of `retroarch.cfg` is a mess? I am interested in creating a RetroArch module to generate `retroarch.cfg` soon™, and on https://github.com/NixOS/nixpkgs/pull/146714 I patched the RetroArch source code to make some configurations immutable (so they always load their defaults). This seems to fix the most issues I found in our old RetroArch derivation, but I may be missing something.

**Comment by bjornfor** [comment]
AFAIR, `retroarch.cfg` is a combination of user configuration and install time configuration:

* Installing assets requires mutating retroarch.cfg. So it basically it leaves an important part of the application to be finalized at runtime -- without this there is no UI!
* The program install prefix is stored there multiple times. It means keeping the file in $HOME breaks after a while as Nix store paths get garbage collected. And for non-NixOS / flatpak, where no paths will be garbage collected, you still cannot share retroarch.cfg across systems due to the install prefix being leaked here.
* Gampad user data is stored there. That's user data, it belongs somewhere in $HOME ($XDG_DATA_DIR?). This is in conflict with the other use of the file, like where the program is installed and where its resource/asset files are.

To sum up, the file should be split in at least two separate files: one read-only part for the application (or simply build in sane defaults[1] and one mutable user configuration part to keep in $HOME.

I think I saw an upstream issue about this.

[1] IIRC, retroarch dumps lots of stuff into retroarch.cfg on startup, which makes version controlling your user data difficult and you end up with outdated configuration over time as your system remembers whatever settings was the default at the time you ran retroarch, not what the defaults are for your current retroarch.

**Comment by thiagokokada** [comment]
> Installing assets requires mutating retroarch.cfg. So it basically it leaves an important part of the application to be finalized at runtime -- without this there is no UI!

Yeah, this sure is annoying. I hope that having a module and having some option like `program.retroarch.theme` or something can fix this issue properly. However, it is not that bad: at least RetroArch correctly points up that the assets are missing at startup and guides the user what needs to be done to fix the missing assets.

> The program install prefix is stored there multiple times. It means keeping the file in $HOME breaks after a while as Nix store paths get garbage collected. And for non-NixOS / flatpak, where no paths will be garbage collected, you still cannot share retroarch.cfg across systems due to the install prefix being leaked he

You're probably referring to those two options:
```
$ cat .config/retroarch/retroarch.cfg | fgrep /nix/store
libretro_directory = "/nix/store/4l3q897h2hwqw382b852j1l0x6dnyn6l-retroarch-1.9.13.2/lib/"
libretro_info_path = "/nix/store/kiycxa1dd6pq9v1rahi8llq53bryj5mm-retroarch-bare-1.9.13.2/share/libretro/info/"
```

After https://github.com/NixOS/nixpkgs/pull/146714/ this shouldn't be an issue anymore. They're essentially read-only options since whatever you set on those options are ignored at RetroArch startup and are overwritten with the (correct) defaults. So no more issues with it pointing to the old values and a GC trigger removing it, broken your configuration until you fix the config file.

> Gampad user data is stored there. That's user data, it belongs somewhere in $HOME ($XDG_DATA_DIR?). This is in conflict with the other use of the file, like where the program is installed and where its resource/asset files are.

I also think having a module `programs.retroarch` with the option to configure anything (including the gamepad) should fix this issue.

> To sum up, the file should be split in at least two separate files: one read-only part for the application (or simply build in sane defaults[1] and one mutable user configuration part to keep in $HOME.
> I think I saw an upstream issue about this.

Yeah, this would be great. The patch on https://github.com/NixOS/nixpkgs/pull/146714 is not big, however it is kinda fragile (this is why I tried to document it as much as possible). Having upstream separating its concerns correctly would be better to avoid unnecessary patching.

**Comment by thiagokokada** [comment]
Closing this since the team was created in #193939.
